src/storage/mvcc/reader/scanner/delta.rs

```diff
@@ -0,0 +1,590 @@
+
+//! A scanner that outputs content in CF between (start_ts,end_ts]
+use std::cmp::Ordering;
+
+use engine::CF_DEFAULT;
+
+use crate::storage::kv::SEEK_BOUND;
+use crate::storage::mvcc::lock::{Lock, LockType};
+use crate::storage::mvcc::write::{Write, WriteRef, WriteType};
+use crate::storage::mvcc::Result;
+use crate::storage::txn::{Result as TxnResult, TxnEntry, TxnEntryScanner};
+use crate::storage::{Cursor, Key, KvPair, Snapshot, Statistics, TimeStamp};
+
+use super::ScannerConfig;
+struct LockEntry {
+    user_key: Key,
+    entry: TxnEntry,
+}
+
+struct LockItem {
+    lock: Lock,
+    key: Vec<u8>,
+    value: Vec<u8>,
+}
+
+struct WriteEntry {
+    user_key: Key,
+    entry: TxnEntry,
+}
+
+pub struct DeltaScanner<S: Snapshot> {
+    cfg: ScannerConfig<S>,
+    // if is none, we needn't scan it
+    lock_cursor: Option<Cursor<S::Iter>>,
+    write_cursor: Cursor<S::Iter>,
+    default_cursor: Option<Cursor<S::Iter>>,
+    /// Is iteration started
+    is_started: bool,
+    statistics: Statistics,
+    // the end ts is cfg.ts
+    begin_ts: TimeStamp,
+    // check and return error if a lock exist
+    err_lock_exist: bool,
+    cache_lock: Option<LockEntry>,
+    cache_write: Option<WriteEntry>,
+}
+
+impl<S: Snapshot> TxnEntryScanner for DeltaScanner<S> {
+    fn next_entry(&mut self) -> TxnResult<Option<TxnEntry>> {
+        Ok(self.read_next()?)
+    }
+    fn take_statistics(&mut self) -> Statistics {
+        std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+}
+
+impl<S: Snapshot> DeltaScanner<S> {
+    pub fn new(
+        cfg: ScannerConfig<S>,
+        lock_cursor: Option<Cursor<S::Iter>>,
+        write_cursor: Cursor<S::Iter>,
+        begin_ts: TimeStamp,
+        err_lock_exist: bool,
+    ) -> DeltaScanner<S> {
+        DeltaScanner {
+            cfg,
+            lock_cursor,
+            write_cursor,
+            statistics: Statistics::default(),
+            default_cursor: None,
+            is_started: false,
+            begin_ts,
+            err_lock_exist,
+            cache_lock: None,
+            cache_write: None,
+        }
+    }
+
+    pub fn read_next(&mut self) -> Result<Option<TxnEntry>> {
+        self.check_and_start()?;
+        if self.lock_cursor.is_none() {
+            let data = self.read_next_commit_item()?;
+            return Ok(data.map(|e| e.entry));
+        }
+        if self.err_lock_exist {
+            return self.read_and_check_lock();
+        }
+        self.read_next_item()
+    }
+
+    fn end_ts(&self) -> TimeStamp {
+        self.cfg.ts
+    }
+
+    fn read_and_check_lock(&mut self) -> Result<Option<TxnEntry>> {
+        let lock_item = self.read_next_lock_item()?;
+        if let Some(lock) = lock_item {
+            let current_user_key = Key::from_encoded_slice(&lock.key);
+            lock.lock.check_ts_conflict(
+                &current_user_key,
+                self.end_ts(),
+                &self.cfg.bypass_locks,
+            )?;
+        }
```

Comment:
This will return the key is locked error before scanning any data, and there may be many valid data before the lock. Is ths behavior ok?